File: server/api/views/nodeinfo.py
from rest_framework.decorators import api_view
from rest_framework.response import Response
from rest_framework.reverse import reverse

from api.models import User
from lightpub.settings import VERSION


def get_total_users():
    # count of local users
    return User.objects.filter(host=None).count()


def _gen_data(version: str):
    return {
        "version": version,
        "software": {
            "name": "lightpub",
            "version": VERSION,
        },
        "protocol": [
            "activitypub",
        ],
        "services": {"inbound": [], "outbound": []},
        "openRegistrations": True,
        "usage": {
            "users": {
                "total": get_total_users(),
            }
        },
        "metadata": {},
    }


@api_view(["GET"])
def version_2_1(request):
    return Response(status=200, data=_gen_data(version="2.1"))


@api_view(["GET"])
def version_2_0(request):
    return Response(status=200, data=_gen_data(version="2.0"))


@api_view(["GET"])
def nodeinfo(request):
    return Response(
        status=200,
        data={
            "links": [
                {
                    "rel": "http://nodeinfo.diaspora.software/ns/schema/2.1",
                    "href": reverse("api:nodeinfo-2.1", request=request),
                },
                {
                    "rel": "http://nodeinfo.diaspora.software/ns/schema/2.0",
                    "href": reverse("api:nodeinfo-2.0", request=request),
                },
            ]
        },
    )

File: src/api/federation/nodeinfo.rs
use actix_web::{get, web, HttpResponse, Responder};
use lightpub_service::{
    services::{note::count_local_notes, user::get_total_users_count},
    ServiceResult,
};
use serde_json::json;

use crate::AppState;

#[get("/.well-known/nodeinfo")]
pub async fn nodeinfo(st: web::Data<AppState>) -> impl Responder {
    let href = st.base_url().join("/nodeinfo/2.1").unwrap();
    let data = json!({
        "links": [
            {
                "rel": "http://nodeinfo.diaspora.software/ns/schema/2.1",
                "href": href
            }
        ]
    });
    HttpResponse::Ok().json(data)
}

#[get("/nodeinfo/2.1")]
pub async fn nodeinfo_2_1(st: web::Data<AppState>) -> ServiceResult<impl Responder> {
    let total_users_count = get_total_users_count(&st.maybe_conn()).await?;
    let total_notes_count = count_local_notes(&st.maybe_conn()).await?;

    let data = json!({
        "openRegistrations": st.is_registration_open(),
        "protocols": [
            "activitypub"
        ],
        "software": {
            "name": "Lightpub",
            "version": env!("CARGO_PKG_VERSION")
        },
        "usage": {
            "users": {
                "total": total_users_count
            },
            "localPosts": total_notes_count,
        },
        "services": {
            "inbound": [],
            "outbound": []
        },
        "metadata": {
            "nodeName": st.nodeinfo().name(),
            "nodeDescription": st.nodeinfo().description(),
        },
        "version": "2.1"
    });

    Ok(HttpResponse::Ok().json(data))
}
